core/dataset_operations.cpp

```diff
@@ -146,38 +183,10 @@ Dataset concatenate(const DatasetConstProxy &a, const DatasetConstProxy &b,
   return result;
 }
 
-template <class Func, class... Args>
-DataArray apply_and_drop_dim(const DataConstProxy &a, Func func, const Dim dim,
-                             Args &&... args) {
-  std::map<Dim, Variable> coords;
-  for (auto && [ d, coord ] : a.coords())
-    if (d != dim)
-      coords.emplace(d, coord);
-
-  std::map<std::string, Variable> labels;
-  for (auto && [ name, label ] : a.labels())
-    if (label.dims().inner() != dim)
-      labels.emplace(name, label);
-
-  std::map<std::string, Variable> attrs;
-  for (auto && [ name, attr ] : a.attrs())
-    if (attr.dims().inner() != dim)
-      attrs.emplace(name, attr);
-
-  return DataArray(func(a.data(), dim, std::forward<Args>(args)...),
-                   std::move(coords), std::move(labels), std::move(attrs));
-}
-
-template <class Func, class... Args>
-Dataset apply_to_items(const DatasetConstProxy &d, Func func, Args &&... args) {
-  Dataset result;
-  for (const auto & [ name, data ] : d)
-    result.setData(name, func(data, std::forward<Args>(args)...));
-  return result;
-}
-
 DataArray sum(const DataConstProxy &a, const Dim dim) {
-  return apply_and_drop_dim(a, [](auto &&... _) { return sum(_...); }, dim);
+  return apply_and_drop_dim(
+      a, [](const auto &d, const auto _dim) { return sum(d.data(), _dim); },
```

Comment:
I find the changed logic a bit odd: `apply_and_drop_dim` now expects a lambda that takes a `DataArray` but returns a `Variable`. I guess you could not use other arguments since `data()` throws for unweighted event data? Maybe `apply_and_drop_dim` can be changed to support both options:

```cpp
template <bool ApplyToData, (existing args)>
auto apply_and_drop_dim_impl(...) {
  if constexpr(ApplyToData)
    func(item.data());
  else
    func(item);
}

template <class... Args>
auto apply_to_data_and_drop_dim(...) {
  return apply_and_drop_dim_impl<true>(std::forward<Args>(args)...);
}
````
and the same with `false` for what you need for `histogram`?